## Duplicated features with Feature-engine

In this notebook, we will identify and remove duplicated features with Feature-engine.

In [1]:
# import libraries

import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline

from feature_engine.selection import DropConstantFeatures, DropDuplicateFeatures

In [2]:
# load data

data = pd.read_csv('../precleaned-datasets/dataset_1.csv')
data.head()

var_1  var_2  var_3  var_4  var_5  var_6  var_7  ...  var_295  var_296  var_297  var_298  var_299     var_300  target
0      0      0    0.0   0.00    0.0      0      0  ...        0        0        0        0      0.0      0.0000       0
1      0      0    0.0   3.00    0.0      0      0  ...        0        0        0        0      0.0      0.0000       0
2      0      0    0.0   5.88    0.0      0      0  ...        3        0        0        0      0.0  67772.7216       0
3      0      0    0.0  14.10    0.0      0      0  ...        0        0        0        0      0.0      0.0000       0
4      0      0    0.0   5.76    0.0      0      0  ...        0        0        0        0      0.0      0.0000       0

[5 rows x 301 columns]

In [3]:
data.shape

(50000, 301)

**Important**

In all feature selection procedures, it is a good practice to select the features by examining only the training set. And this is to avoid overfit.

In [4]:
# separate dataset into train and test

X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels=['target'], axis=1), # drop the target
    data['target'], # just the target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 300), (15000, 300))

## Remove constant and quasi-constant

In [5]:
# remove constant and quasi-constant features first:
# we use Feature-engine for this

sel = DropConstantFeatures(tol=0.998, variables=None, missing_values='raise')

sel.fit(X_train)

DropConstantFeatures(tol=0.998,
                     variables=['var_1', 'var_2', 'var_3', 'var_4', 'var_5',
                                'var_6', 'var_7', 'var_8', 'var_9', 'var_10',
                                'var_11', 'var_12', 'var_13', 'var_14',
                                'var_15', 'var_16', 'var_17', 'var_18',
                                'var_19', 'var_20', 'var_21', 'var_22',
                                'var_23', 'var_24', 'var_25', 'var_26',
                                'var_27', 'var_28', 'var_29', 'var_30', ...])

In [6]:
sel.features_to_drop_

['var_1',
 'var_2',
 'var_3',
 'var_6',
 'var_7',
 'var_9',
 'var_10',
 'var_11',
 'var_12',
 'var_14',
 'var_16',
 'var_20',
 'var_23',
 'var_24',
 'var_28',
 'var_32',
 'var_33',
 'var_34',
 'var_36',
 'var_39',
 'var_40',
 'var_42',
 'var_43',
 'var_44',
 'var_45',
 'var_48',
 'var_53',
 'var_56',
 'var_59',
 'var_60',
 'var_61',
 'var_65',
 'var_66',
 'var_67',
 'var_69',
 'var_71',
 'var_72',
 'var_73',
 'var_77',
 'var_78',
 'var_80',
 'var_81',
 'var_87',
 'var_89',
 'var_90',
 'var_92',
 'var_95',
 'var_97',
 'var_98',
 'var_99',
 'var_102',
 'var_104',
 'var_106',
 'var_111',
 'var_112',
 'var_113',
 'var_115',
 'var_116',
 'var_120',
 'var_122',
 'var_124',
 'var_125',
 'var_126',
 'var_127',
 'var_129',
 'var_130',
 'var_133',
 'var_135',
 'var_136',
 'var_138',
 'var_141',
 'var_142',
 'var_146',
 'var_149',
 'var_150',
 'var_151',
 'var_153',
 'var_158',
 'var_159',
 'var_167',
 'var_170',
 'var_171',
 'var_178',
 'var_180',
 'var_182',
 'var_183',
 'var_184',
 'var_187',


In [7]:
# remove the quasi-constant features

X_train = sel.transform(X_train)
X_test = sel.transform(X_test)

X_train.shape, X_test.shape

((35000, 158), (15000, 158))

- Now we can see, we are left with only 158 features after removing the Constant and Quasi Constant features

## Remove duplicated features

In [8]:
# set up the selector
sel = DropDuplicateFeatures(variables=None, missing_values='raise')

# find the duplicate features, this might take a while
sel.fit(X_train)

DropDuplicateFeatures(missing_values='raise',
                      variables=['var_4', 'var_5', 'var_8', 'var_13', 'var_15',
                                 'var_17', 'var_18', 'var_19', 'var_21',
                                 'var_22', 'var_25', 'var_26', 'var_27',
                                 'var_29', 'var_30', 'var_31', 'var_35',
                                 'var_37', 'var_38', 'var_41', 'var_46',
                                 'var_47', 'var_49', 'var_50', 'var_51',
                                 'var_52', 'var_54', 'var_55', 'var_57',
                                 'var_58', ...])

In [9]:
sel.duplicated_feature_sets_

[{'var_148', 'var_37'},
 {'var_199', 'var_84'},
 {'var_143', 'var_296'},
 {'var_177', 'var_250'},
 {'var_226', 'var_232'},
 {'var_229', 'var_269'}]

In [10]:
sel.features_to_drop_

{'var_148', 'var_199', 'var_232', 'var_250', 'var_269', 'var_296'}

In [11]:
# remove the duplicated features

X_train = sel.transform(X_train)
X_test = sel.transform(X_test)

X_train.shape, X_test.shape

((35000, 152), (15000, 152))

- Now the duplicated features are also removed, so left with only 152 features, almost half of the original features set

## Stack Feature selection in a Pipeline

We can perform both steps together by setting up the transformers within a pipeline.

In [12]:
# separate dataset into train and test

X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels=['target'], axis=1), # drop the target
    data['target'], # just the target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 300), (15000, 300))

In [13]:
pipe = Pipeline(steps = [
    ('constant', DropConstantFeatures(tol=0.998, variables=None, missing_values='raise')),
    ('duplicates', DropDuplicateFeatures(variables=None, missing_values='raise'))
])

pipe.fit(X_train)

Pipeline(steps=[('constant',
                 DropConstantFeatures(tol=0.998,
                                      variables=['var_1', 'var_2', 'var_3',
                                                 'var_4', 'var_5', 'var_6',
                                                 'var_7', 'var_8', 'var_9',
                                                 'var_10', 'var_11', 'var_12',
                                                 'var_13', 'var_14', 'var_15',
                                                 'var_16', 'var_17', 'var_18',
                                                 'var_19', 'var_20', 'var_21',
                                                 'var_22', 'var_23', 'var_24',
                                                 'var_25', 'var_26', 'var_27',
                                                 'var_28', 'var_29', 'var_30', ...])),
                ('duplicates',
                 DropDuplicateFeatures(missing_values='raise',
                                       

In [15]:
pipe.named_steps['duplicates'].features_to_drop_

{'var_148', 'var_199', 'var_232', 'var_250', 'var_269', 'var_296'}

In [16]:
# remove features

X_train = pipe.transform(X_train)
X_test = pipe.transform(X_test)

X_train.shape, X_test.shape

((35000, 152), (15000, 152))

In [17]:
# we can navigate the pipeline transformers

len(pipe.named_steps['constant'].features_to_drop_)

142